### DISCUSSION TITLE
Vitest in browser, for in browser testing

### DISCUSSION THREAD

**Discussion by gqio** [vitest-dev/vitest#23] [General]
Hello @patak-js ,  @antfu ,  congrats great work around jest, it's refreshing.
I tried to integrate vitest inside our browser testing implementation, I had some issues 😦 Here is some issues I had during my integration:
- I think the run is too tied to fs and others node-only stuff. I like the idea of the runner and suites, but for being able to use only the runner it requires a lot of alias inside our build and mocking.
- by default the lib has narrow exports and it's blocking (at least me :)) to use internal stuff like collectFiles and runFiles, I managed to remove them locally but won't work by default.
- To me the DefaultReporter should not be part of the run module because it's linked to a lot to node environement.
- I wish I could hook/replace the snapshotManager 😍  (via config) like the reporter but I need different access to files than the node fs layer. And I think the exposed api could have been just matchSnapshot, toSnapshot, saveSnapshot, or something like this. I don't have a clear idea yet. Because for now I cannot use it to publish/save somewhere a snapshot
- I think also in the TaskResult api it lacks a duration or maybe it's me and I couldn't find it. As a suggestion I liked also what mocha did on timing by defining acceptable running timing (fast/medium/slow) being able to categorize them but for sure could be done with the reporter api.
https://imgur.com/a/jex9gTJ
and here is the snippet of code I use now for being able to run vitest in the browser https://gist.github.com/gqio/404001da5333d86f408e79ba19e33c8f

Sorry for the reporter not being very neat but at least it adapts my actual front report api and the vitest runner api 🙂 
FTR I commented in the run module those lines to avoid bundling issue on my side
```
// import fg from 'fast-glob';
// import { setupChai } from '../integrations/chai/setup';
// import { DefaultReporter } from '../reporters/default';
// import { getSnapshotManager } from '../integrations/chai/snapshot';
```

Maybe splitting core feature and runtime could fit the need. 

Wdyt ?

**Comment by antfu** [comment]
Hi Gregory, thanks for bringing this up. However, Vitest is designed to be running on Node only. The whole idea of this project is to run Vite on the Node side so you don't need to fire up the entire browser to run the test. I think for browser testing, Cypress or Web test runner could better fit the job (I'd personally recommend Vitest + Cypress, which both are using the same Vite engine to cover both units and e2e testing on Node and browser). Or, can you share more about what's your usage, and maybe we could see if it's could fit in Vitest's scope?

To me personally snapshotting is an important feature that I found testing frameworks other than Jest are missing. In addition to having everything in the box, that's kinda what I love using Jest about. So I kinda targeting Vitest as an alternative to Jest with less coupled deps while making them work nicely together.

> Expose more APIs like `collectFiles` and `runFile`

It's easy for us to expose them. But the important thing is that Vitest has to be running on Vite interpreted env ([vite-node](https://github.com/antfu/vite-node)) but not native Node.js. I haven't thought about the programmatic API yet, but maybe we could in the future. 

> Swapable snapshotManager

We could do that, but I am not sure if `jest-snapshot` is friendly on browsers, otherwise, you will need to implement it from scratch.

> I think also in the TaskResult api it lacks a duration

Yes. That's something left to do, PR welcome :)

**Comment by taye** [comment]
Thanks @antfu and other contributors for vitest. It's a great project.

I made something sort of similar (which I was actually going to call vitest :sweat_smile:) that patched jest to to allow a custom runner that runs the tests with puppeteer https://github.com/taye/vijest. It uses synchronous XHR in the browser to communicate with the test runner and have snapshots and reading from the FS with the usual jest APIs.

I'd be very happy to try adding something similar to vitest as I think for some cases:

 - Cypress is a bit too much
 - with jsdom and happy-dom, many DOM APIs that media and interaction-focused apps and libraries need are missing
 - it's nice to be able to switch from jsdom/happy-dom to a headed browser to watch and debug what's going on in the tests

**Reply by sheremet-va** [reply]
You can look into https://github.com/vitest-dev/vitest/pull/2517 and https://github.com/vitest-dev/vitest/tree/main/packages/browser

**Reply by mercmobily** [reply]
Wow this is quite huge... Does this sit in the same use case as Playright?

**Reply by sheremet-va** [reply]
The closest use case would be Playwright Component Testing. Vitest doesn't plan to provide a lot of utilities out of the box (like network interception). We also do not plan to support e2e testing as we believe Playwright already provides a very good experience. Vitest itself uses playwright (or webdriverio) for its browser mode.

I'd recommend thinking about Vitest browser mode as an alternative to jsdom - you should be able to just make a switch without much changes.

**Comment by kinthaiofficial** [comment]
Browser-mode testing in Vitest is one of the most valuable features for testing components that need real browser APIs — things that jsdom gets wrong or doesn't support at all.

The key use cases where browser mode wins over jsdom:

```typescript
// These things break or behave differently in jsdom:

// 1. ResizeObserver, IntersectionObserver
// 2. Canvas/WebGL
// 3. Web Animations API
// 4. CSS custom properties computed values
// 5. Native form validation
// 6. File API with actual file objects
// 7. Clipboard API
// 8. Service Workers

// In browser mode, these all work natively:
test('ResizeObserver fires on element resize', async () => {
  const el = document.createElement('div');
  document.body.appendChild(el);
  
  let observed = false;
  const observer = new ResizeObserver(() => { observed = true; });
  observer.observe(el);
  
  el.style.width = '200px';
  await new Promise(r => requestAnimationFrame(r));
  
  expect(observed).toBe(true);  // Works in browser, fails in jsdom
});
```

**The setup**:
```typescript
// vitest.config.ts
export default defineConfig({
  test: {
    browser: {
      enabled: true,
      name: 'chromium',  // or 'firefox', 'webkit'
      provider: 'playwright',
    },
  },
});
```

**Trade-off**: Browser mode is slower than jsdom (actual browser startup) and requires Playwright/WebdriverIO. The pattern I use: jsdom for pure logic/hooks, browser mode for component tests that need real browser APIs. The `--browser` flag makes it easy to run just browser tests.

What's the specific API you're hitting that jsdom gets wrong? Happy to share how browser mode handles it.
